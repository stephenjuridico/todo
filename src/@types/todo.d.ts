import {RouteProp} from '@react-navigation/native';
import {Priority, SortBy, Status} from '../constants/constants';
import type {NativeStackNavigationProp} from '@react-navigation/native-stack';

export interface ITodo {
  id: string;
  title: string;
  description: string;
  status: Status;
  priority: Priority;
  createdAt?: string;
  updatedAt?: string;
}

export interface IFilters {
  status: string;
  name: SortBy;
  priority: SortBy;
}

export interface Response {
  message: string;
  item?: ITodo;
  error?: any;
}

export type TodoContextType = {
  data: ITodo[];
  mapped: MappedTodoType;
  filters: IFilters;
  isLoading: boolean;
  createTodo: (todo: ITodo) => Promise<Response>;
  updateTodo: (todo: ITodo) => Promise<Response>;
  updateStatus: (
    id: string,
    status: Status,
    showToast?: boolean,
  ) => Promise<Response>;
  setFilters: (filters: IFilters) => void;
};

export type MappedTodoType = {[key: string]: {count: number; items: ITodo[]}};

export type RootStackParamList = {
  Main: undefined;
  CreateTask: {item: ITodo} | undefined;
};

export type RootRouteProps<RouteName extends keyof RootStackParamList> =
  RouteProp<RootStackParamList, RouteName>;

export type MainScreenNavigationProp = NativeStackNavigationProp<
  RootStackParamList,
  'Main'
>;
