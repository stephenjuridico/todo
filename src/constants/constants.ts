export enum Status {
  Pending = 'pending',
  Completed = 'completed',
  Archived = 'archived',
}

export const StatusList = Object.entries(Status).map(([k, v]) => ({
  label: k,
  value: v,
}));

export enum Priority {
  High = 3,
  Medium = 2,
  Low = 1,
  None = 0,
}

export const PriorityColorCode = {
  3: 'red',
  2: 'amber',
  1: 'green',
  0: 'gray',
};

export const PriorityList = Object.entries(Priority)
  .filter(([k, _]) => isNaN(Number(k)))
  .map(([key, value]) => ({
    label: key,
    value,
    color: PriorityColorCode[value as Priority],
  }));

export enum SortBy {
  ASC = 'ascend',
  DESC = 'descend',
}
