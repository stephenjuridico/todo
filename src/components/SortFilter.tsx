import React, {forwardRef, useImperativeHandle} from 'react';
import {
  Box,
  HStack,
  Icon,
  Text,
  useDisclose,
  IconButton,
  Actionsheet,
} from 'native-base';
import Feather from 'react-native-vector-icons/Feather';
import {useTodo} from '../context/TodoContext';
import {SortBy} from '../constants/constants';

const SortFilter = forwardRef((props, ref) => {
  const {isOpen, onClose, onToggle} = useDisclose();
  const {filters, setFilters} = useTodo();

  useImperativeHandle(
    ref,
    () => {
      return {
        toggle() {
          onToggle();
        },
      };
    },
    [onToggle],
  );

  return (
    <Actionsheet isOpen={isOpen} onClose={onClose}>
      <Actionsheet.Content>
        <Box w="100%" h={60} px={4} justifyContent="center">
          <Text bold fontSize="16">
            Sort By
          </Text>
          <HStack alignItems={'center'} justifyContent={'space-between'}>
            <Text>Title</Text>
            <IconButton
              borderRadius={'full'}
              testID="btnSortByName"
              onPress={() =>
                setFilters({
                  ...filters,
                  name: filters.name === SortBy.ASC ? SortBy.DESC : SortBy.ASC,
                })
              }
              icon={
                <Icon
                  as={Feather}
                  name={filters.name === SortBy.ASC ? 'arrow-up' : 'arrow-down'}
                />
              }
            />
          </HStack>
          <HStack alignItems={'center'} justifyContent={'space-between'}>
            <Text>Priority</Text>
            <IconButton
              testID="btnSortByPriority"
              borderRadius={'full'}
              onPress={() =>
                setFilters({
                  ...filters,
                  priority:
                    filters.priority === SortBy.ASC ? SortBy.DESC : SortBy.ASC,
                })
              }
              icon={
                <Icon
                  as={Feather}
                  name={
                    filters.priority === SortBy.ASC ? 'arrow-up' : 'arrow-down'
                  }
                />
              }
            />
          </HStack>
        </Box>
      </Actionsheet.Content>
    </Actionsheet>
  );
});

export default SortFilter;
