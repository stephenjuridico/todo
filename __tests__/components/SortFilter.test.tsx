import 'react-native';
import React from 'react';
import SortFilter from '../../src/components/SortFilter';
import {
  act,
  fireEvent,
  render,
  renderHook,
  screen,
  waitFor,
} from '../../src/test/test-utils';
import {
  TodoProvider,
  useTodo,
  initialMappedValue,
  initialFiltersValue,
} from '../../src/context/TodoContext';
import {TodoContextType} from '../../src/@types/todo';
import {SortBy} from '../../src/constants/constants';

describe('SortFilter', () => {
  const contextValues = {
    data: [],
    mapped: initialMappedValue,
    isLoading: false,
    createTodo: jest.fn(),
    updateTodo: jest.fn(),
    updateStatus: jest.fn(),
    filters: initialFiltersValue,
    setFilters: jest.fn(),
  } as TodoContextType;

  const renderSortFilter = () => {
    render(
      <TodoProvider value={contextValues}>
        <SortFilter />
      </TodoProvider>,
    );
  };

  it('renders correctly', async () => {
    renderSortFilter();
    await waitFor(() => {
      expect(screen.toJSON()).toMatchSnapshot();
    });
  });

  // it('renders correctly and displays default filters', () => {
  //   renderSortFilter();
  //   const btnSortByName = screen.getByTestId('btnSortByName');
  //   const btnSortByPriority = screen.getByTestId('btnSortByPriority');
  //   fireEvent.press(btnSortByName);
  //   fireEvent.press(btnSortByPriority);
  //   expect(contextValues.filters.priority).toBe(SortBy.DESC);
  //   expect(contextValues.filters.name).toBe(SortBy.ASC);
  // });

  // it('renders correctly and clicks sort by name', async () => {
  //   renderSortFilter();
  //   const btnSortByName = screen.getByTestId('btnSortByName');
  //   fireEvent.press(btnSortByName);
  //   expect(contextValues.setFilters).toBeCalled();
  // });

  // it('renders correctly and clicks sort by priority', async () => {
  //   renderSortFilter();
  //   const btnSortByPriority = screen.getByTestId('btnSortByPriority');
  //   fireEvent.press(btnSortByPriority);
  //   expect(contextValues.setFilters).toBeCalled();
  // });
});
