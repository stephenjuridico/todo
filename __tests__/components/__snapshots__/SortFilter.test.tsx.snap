// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortFilter renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      [
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        {
          "display": "flex",
          "overflow": "hidden",
        },
      ]
    }
  >
    <View
      collapsable={false}
      pointerEvents="none"
      style={
        {
          "bottom": 0,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "transform": [
            {
              "translateY": 0,
            },
            {
              "translateX": 0,
            },
            {
              "scale": 1,
            },
            {
              "scaleX": 1,
            },
            {
              "scaleY": 1,
            },
            {
              "rotate": "0deg",
            },
          ],
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={false}
        collapsable={false}
        dataSet={{}}
        focusable={true}
        importantForAccessibility="no"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "rgb(0, 0, 0)",
            "bottom": 0,
            "left": 0,
            "opacity": 0.3,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      />
    </View>
    <View
      collapsable={false}
      pointerEvents="none"
      style={
        {
          "bottom": 0,
          "height": "100%",
          "left": 0,
          "opacity": 1,
          "position": "absolute",
          "right": 0,
          "transform": [
            {
              "translateY": 0,
            },
            {
              "translateX": 0,
            },
            {
              "scale": 1,
            },
            {
              "scaleX": 1,
            },
            {
              "scaleY": 1,
            },
            {
              "rotate": "0deg",
            },
          ],
        }
      }
    >
      <View
        dataSet={{}}
        onLayout={[Function]}
        pointerEvents="box-none"
        style={
          {
            "height": "100%",
            "opacity": 0,
          }
        }
      >
        <View
          closeOnOverlayClick={true}
          dataSet={{}}
          pointerEvents="box-none"
          style={
            {
              "alignItems": "center",
              "height": "100%",
              "justifyContent": "flex-end",
              "width": "100%",
            }
          }
        >
          <View
            collapsable={false}
            onLayout={[Function]}
            pointerEvents="box-none"
            style={
              {
                "transform": [
                  {
                    "translateY": 0,
                  },
                ],
                "width": "100%",
              }
            }
          >
            <View
              collapsable={false}
              dataSet={{}}
              onMoveShouldSetResponder={[Function]}
              onMoveShouldSetResponderCapture={[Function]}
              onResponderEnd={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderReject={[Function]}
              onResponderRelease={[Function]}
              onResponderStart={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              onStartShouldSetResponderCapture={[Function]}
              style={
                {
                  "paddingBottom": 8,
                  "paddingTop": 8,
                }
              }
            />
            <View
              accessibilityViewIsModal={true}
              aria-modal="true"
              dataSet={{}}
              onAccessibilityEscape={[Function]}
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "#fafafa",
                  "borderRadius": 0,
                  "borderTopLeftRadius": 20,
                  "borderTopRightRadius": 20,
                  "elevation": 2,
                  "maxHeight": 1184,
                  "overflow": "hidden",
                  "paddingBottom": 8,
                  "paddingLeft": 8,
                  "paddingRight": 8,
                  "paddingTop": 8,
                  "shadowColor": "#000000",
                  "shadowOffset": {
                    "height": 1,
                    "width": 0,
                  },
                  "shadowOpacity": 0.2,
                  "shadowRadius": 1.41,
                  "width": "100%",
                }
              }
            >
              <View
                collapsable={false}
                dataSet={{}}
                onMoveShouldSetResponder={[Function]}
                onMoveShouldSetResponderCapture={[Function]}
                onResponderEnd={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderReject={[Function]}
                onResponderRelease={[Function]}
                onResponderStart={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                onStartShouldSetResponderCapture={[Function]}
                style={
                  {
                    "alignItems": "center",
                    "marginTop": -8,
                    "paddingBottom": 12,
                    "paddingTop": 12,
                    "width": "100%",
                  }
                }
              >
                <View
                  dataSet={{}}
                  style={
                    {
                      "backgroundColor": "#737373",
                      "borderRadius": 2,
                      "height": 4,
                      "width": 40,
                    }
                  }
                />
              </View>
              <View
                dataSet={{}}
                style={
                  {
                    "height": 60,
                    "justifyContent": "center",
                    "paddingLeft": 16,
                    "paddingRight": 16,
                    "width": "100%",
                  }
                }
              >
                <Text
                  dataSet={{}}
                  style={
                    {
                      "backgroundColor": undefined,
                      "color": "#171717",
                      "fontFamily": undefined,
                      "fontSize": 16,
                      "fontStyle": "normal",
                      "fontWeight": "700",
                      "letterSpacing": 0,
                      "lineHeight": 24,
                      "textDecorationLine": undefined,
                    }
                  }
                >
                  Sort By
                </Text>
                <View
                  dataSet={{}}
                  style={
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "space-between",
                    }
                  }
                >
                  <Text
                    dataSet={{}}
                    style={
                      {
                        "backgroundColor": undefined,
                        "color": "#171717",
                        "fontFamily": undefined,
                        "fontSize": 14,
                        "fontStyle": "normal",
                        "fontWeight": "400",
                        "letterSpacing": 0,
                        "lineHeight": 21,
                        "textDecorationLine": undefined,
                      }
                    }
                  >
                    Title
                  </Text>
                  <View
                    dataSet={{}}
                    style={{}}
                  />
                  <View
                    accessibilityRole="button"
                    accessibilityState={
                      {
                        "busy": undefined,
                        "checked": undefined,
                        "disabled": undefined,
                        "expanded": undefined,
                        "selected": undefined,
                      }
                    }
                    accessibilityValue={
                      {
                        "max": undefined,
                        "min": undefined,
                        "now": undefined,
                        "text": undefined,
                      }
                    }
                    accessible={true}
                    collapsable={false}
                    colorScheme="primary"
                    dataSet={{}}
                    focusable={true}
                    onBlur={[Function]}
                    onClick={[Function]}
                    onFocus={[Function]}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      {
                        "alignItems": "center",
                        "borderRadius": 9999,
                        "flexDirection": "row",
                        "justifyContent": "center",
                        "paddingBottom": 10,
                        "paddingLeft": 10,
                        "paddingRight": 10,
                        "paddingTop": 10,
                      }
                    }
                    testID="btnSortByName"
                  >
                    <Text
                      allowFontScaling={false}
                      dataSet={{}}
                      selectable={false}
                      style={
                        [
                          {
                            "color": undefined,
                            "fontSize": 12,
                          },
                          {
                            "color": "#e6b200",
                            "fontSize": 20,
                            "height": 20,
                            "lineHeight": 20,
                            "width": 20,
                          },
                          {
                            "fontFamily": "Feather",
                            "fontStyle": "normal",
                            "fontWeight": "normal",
                          },
                          {},
                        ]
                      }
                    >
                      
                    </Text>
                  </View>
                </View>
                <View
                  dataSet={{}}
                  style={
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "space-between",
                    }
                  }
                >
                  <Text
                    dataSet={{}}
                    style={
                      {
                        "backgroundColor": undefined,
                        "color": "#171717",
                        "fontFamily": undefined,
                        "fontSize": 14,
                        "fontStyle": "normal",
                        "fontWeight": "400",
                        "letterSpacing": 0,
                        "lineHeight": 21,
                        "textDecorationLine": undefined,
                      }
                    }
                  >
                    Priority
                  </Text>
                  <View
                    dataSet={{}}
                    style={{}}
                  />
                  <View
                    accessibilityRole="button"
                    accessibilityState={
                      {
                        "busy": undefined,
                        "checked": undefined,
                        "disabled": undefined,
                        "expanded": undefined,
                        "selected": undefined,
                      }
                    }
                    accessibilityValue={
                      {
                        "max": undefined,
                        "min": undefined,
                        "now": undefined,
                        "text": undefined,
                      }
                    }
                    accessible={true}
                    collapsable={false}
                    colorScheme="primary"
                    dataSet={{}}
                    focusable={true}
                    onBlur={[Function]}
                    onClick={[Function]}
                    onFocus={[Function]}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      {
                        "alignItems": "center",
                        "borderRadius": 9999,
                        "flexDirection": "row",
                        "justifyContent": "center",
                        "paddingBottom": 10,
                        "paddingLeft": 10,
                        "paddingRight": 10,
                        "paddingTop": 10,
                      }
                    }
                    testID="btnSortByPriority"
                  >
                    <Text
                      allowFontScaling={false}
                      dataSet={{}}
                      selectable={false}
                      style={
                        [
                          {
                            "color": undefined,
                            "fontSize": 12,
                          },
                          {
                            "color": "#e6b200",
                            "fontSize": 20,
                            "height": 20,
                            "lineHeight": 20,
                            "width": 20,
                          },
                          {
                            "fontFamily": "Feather",
                            "fontStyle": "normal",
                            "fontWeight": "normal",
                          },
                          {},
                        ]
                      }
                    >
                      
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
